### ISSUE TITLE
game warns if a .dpk contains a folder named "patches"

### ISSUE THREAD

**Issue by bmorel** [Unvanquished/Unvanquished#2611]
For simplicity's sake (and licence conforming), I want the dpk files containing the .nexe to contain the patches, but when I add a folder named `patches` in there, the game emits a warning.

Is there is a way to store that kind of data without annoying warnings?

**Comment by DolceTriade** [comment]
What's the warning?

**Comment by bmorel** [comment]
`Warn: [FS] Invalid filename 'patches/0010-make-tagging-duration-configurable-and-default-to-0..patch' in pak [...]`

The double dot is probably a way for git to shorten a bit the commit title when it's longer than 52 characters.

**Comment by illwieckz** [comment]
From `Daemon/src/common/FileSystem.cpp`:

```c++
namespace Path {

bool IsValid(Str::StringRef path, bool allowDir)
{
		// …

		// Disallow 2 consecutive / or .
		if ((c == '/' || c == '.') && (prev == '/' || prev == '.'))
			return false;
```

**Comment by bmorel** [comment]
Is there is a reason for that specific check? This is the kind of things that scripts can easily generate, and scripts prevents manual errors. I don't want to have the go through git's (or other tool) generated files just to remove double / or .

**Comment by illwieckz** [comment]
I guess there is, the engine has to split paths on dot, to check for extensions for example, and especially to look for multiple image or sound files for the same basename whatever the extension. We may have made it more permissive as if I'm right, it no longer split on first dot, but one has to check nothing break if removing this check.

**Comment by illwieckz** [comment]
Also mappers may want this check.

**Comment by bmorel** [comment]
yes, it's "0.5" and the shortening happens right at the worst place lol

**Comment by illwieckz** [comment]
You can do that:

```sh
git format-patch -o patches --filename-max-length $((63 - 8)) --suffix '_.patch' <ref>
```

By enforcing a `_` prefix to `.patch` you know there will never be any consecutive dot because of the extension being written at the exact same place following a dot.

Also many parts of the engine assume paths are 64 characters long including the null-terminator, we may add a check for this in the future. `$((63 - 8))` is to substract the length of `patches/`.

**Comment by bmorel** [comment]
Still, what is the reason behind this limitation? The size one looks technical so ok, but the double glyph one?

**Comment by illwieckz** [comment]
> the engine has to split paths on dot, to check for extensions for example, and especially to look for multiple image or sound files for the same basename whatever the extension. 
> -- https://github.com/Unvanquished/Unvanquished/issues/2611#issuecomment-1501771814

It may happen that the engine is more permissive than before, but there can be many historical reasons why double dot is not allowed.

In fact in the past it was even not good to use a dot in a basename (even if not consecutive) like `textures/image.blend.jpg`, though I already removed this limitation.

**Comment by illwieckz** [comment]
Maybe the fix I implemented is enough to also allow `textures/image.blend.jpg` also makes possible to use double dot, one has to check there is no other historical limitation preventing the use of a double dot.

**Comment by bmorel** [comment]
would be nice to remove that, it really feels like artificial limitation
